plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs

```diff
@@ -146,92 +178,82 @@ gameTokenVal =
-guess ::
-    (WalletAPI m, WalletDiagnostics m)
-    => String
-    -- ^ The guess
-    -> String
-    -- ^ A new secret
-    -> Value
-    -- ^ How much ada to take out
-    -> Value
-    -- ^ How much to put back into the contract
-    -> m ()
-guess gss new keepVal restVal = do
-
-    let guessedSecret = ClearString (C.pack gss)
-        newSecret = HashedString (sha2_256 (C.pack new))
-        input = Guess guessedSecret newSecret
-        newState = Locked gameToken newSecret
-        ds = DataScript $ PlutusTx.toData newState
-        redeemer = RedeemerScript $ PlutusTx.toData input
-    ins <- WAPI.spendScriptOutputs (Scripts.validatorScript scriptInstance) redeemer
-    ownOutput <- WAPI.ownPubKeyTxOut (keepVal <> gameTokenVal)
+-- | The @"guess"@ endpoint.
+guess :: AsContractError e => Contract GameStateMachineSchema e ()
+guess = do
+    GuessArgs{guessArgsOldSecret,guessArgsNewSecret, guessArgsValueLocked} <- endpoint @"guess"
+
+    let guessedSecret = ClearString (C.pack guessArgsOldSecret)
+        newSecret     = HashedString (sha2_256 (C.pack guessArgsNewSecret))
+        input         = Guess guessedSecret newSecret
+        newState      = Locked gameToken newSecret
+        ds            = DataScript $ PlutusTx.toData newState
+        redeemer      = RedeemerScript $ PlutusTx.toData input
+        addr          = Scripts.scriptAddress scriptInstance
+
+    -- We need to know the unspent outputs at the contract address, and a public
+    -- key that we can send the token to. 
+    utxo <- utxoAt addr
+    pubKey <- ownPubKey
+
+    -- The transaction has three parts which are joined together using the 
+    -- 'Haskell.<>' semigroup operator.
+    let tx = 
+             -- part 1: Spend the outputs that are currently at the address
+            Tx.collectFromScript
+                utxo 
+                (Scripts.validatorScript scriptInstance)
+                redeemer
+
+             -- part 2: Lock the remaining funds in the contract with the new 
+             -- state 'ds'
+             Haskell.<> Tx.payToScript guessArgsValueLocked addr ds 
+
+             -- part 3: Pay the token to our own public key (this is a no-op on 
+             -- the wallet's balance but it ensures that the token is present 
+             -- in the transaction)
+             Haskell.<> Tx.payToPubKey gameTokenVal pubKey 
+    void $ writeTxSuccess tx
+
+lock :: AsContractError e => Contract GameStateMachineSchema e ()
+lock = do
+    LockArgs{lockArgsSecret, lockArgsValue} <- endpoint @"lock"
+    let secret            = HashedString (sha2_256 (C.pack lockArgsSecret))
+        addr              = Scripts.scriptAddress scriptInstance
+        initialDataScript = DataScript $ PlutusTx.toData (Initialised secret)
+    
+    -- 1. Create a transaction output with the value and the secret and
+    --    wait until it is confirmed.
+    _ <- both (writeTxSuccess $ Tx.payToScript lockArgsValue addr initialDataScript) (fundsAtAddressGt addr zero)
```

Comment:
Why do we need this? Does `writeTxSuccess` not wait until the tx is confirmed?

Reply:
the "success" in `writeTxSuccess` refers to the wallet successfully balancing and signing the transaction, and submitting it to the chain. It doesn't wait until the tx is confirmed. (I think we might need an `awaitTxConfirmed` effect for that)

Reply:
Yeah, and that should be more reliable than the awkward thing we do with tracking the amount at the address currently.